File: 11.py
from copy import deepcopy
import math
import numpy as np
from collections import deque
from itertools import product

ls = []
for s in open('11.in','r').readlines():
    x = s.rstrip('\n')
    ls.append(x)

def one(lin):
    changeflag = True
    cidc = 0
    nl = []
    ls = deepcopy(lin)
    while changeflag:
        cidc += 1
        print(cidc)
        changeflag = False
        ir = 0
        ic = 0
        # print(ls)
        nl = []
        for row in ls:
            r = ""
            ic = 0
            for col in row:
                # if ic==91: continue
                # if ir==95: continue
                if (col=="L"):
                    count = 0
                    for dc,dr in product(range(-1,2,1),range(-1,2,1)):
                        if (ic+dc)==len(row): continue
                        if (ir+dr)==len(ls): continue
                        if (ir+dr)==-1: continue
                        if (ic+dc)==-1: continue
                        if (dc==dr==0): continue
                        if ls[ir+dr][ic+dc] == "#":
                            count+=1
                    if (count==0):
                        r += "#"
                        changeflag = True
                    else:
                        r += "L"
                elif (col=="#"):
                    count = 0
                    for dc,dr in product(range(-1,2,1),range(-1,2,1)):
                        if (ic+dc)==len(row): continue
                        if (ir+dr)==len(ls): continue
                        if (ir+dr)==-1: continue
                        if (ic+dc)==-1: continue
                        if (dc==dr==0): continue
                        if ls[ir+dr][ic+dc]=="#":
                            count+=1
                    if (count>=4):
                        r += "L"
                        changeflag = True
                    else:
                        r += "#"
                else:
                    r += ls[ir][ic]
                ic += 1
            # print(r)
            nl.append(r)
            ir +=1
        ls = deepcopy(nl)
        # print(nl)
    cc = 0
    for row in nl:
        for col in row:
            if col=="#":
                cc+=1
    return cc
    
def two(lin):
    changeflag = True
    cidc = 0
    nl = []
    ls = deepcopy(lin)
    while changeflag:
        cidc += 1
        print(cidc)
        changeflag = False
        ir = 0
        ic = 0
        # print(ls)
        nl = []
        for row in ls:
            r = ""
            ic = 0
            for col in row:
                # if ic==91: continue
                # if ir==95: continue
                if (col=="L"):
                    count = 0
                    for dc,dr in product(range(-1,2,1),range(-1,2,1)):
                        i = 1
                        foundflag = False
                        while not foundflag and i < 95:
                            if (i>1):
                                dc += np.sign(dc)
                                dr += np.sign(dr)
                            i+=1
                            if (ic+dc)>=len(row): break
                            if (ir+dr)>=len(ls): break
                            if (ir+dr)<0: break
                            if (ic+dc)<0: break
                            if (dc==dr==0): break
                            if ls[ir+dr][ic+dc] == "#":
                                count+=1
                            if ls[ir+dr][ic+dc] in {"#","L"}:
                                foundflag=True
                    if (count==0):
                        r += "#"
                        changeflag = True
                    else:
                        r += "L"
                elif (col=="#"):
                    count = 0
                    for dc,dr in product(range(-1,2,1),range(-1,2,1)):
                        i = 0
                        foundflag = False
                        while not foundflag and i < 95:
                            i+=1
                            if (i>1):
                                dc += np.sign(dc)
                                dr += np.sign(dr)
                            if (ic+dc)>=len(row): break
                            if (ir+dr)>=len(ls): break
                            if (ir+dr)<0: break
                            if (ic+dc)<0: break
                            if (dc==dr==0): break
                            if ls[ir+dr][ic+dc] == "#":
                                count+=1
                            if ls[ir+dr][ic+dc] in {"#","L"}:
                                foundflag=True
                    if (count>=5):
                        r += "L"
                        changeflag = True
                    else:
                        r += "#"
                else:
                    r += ls[ir][ic]
                ic += 1
            # print(r)
            nl.append(r)
            ir +=1
        ls = deepcopy(nl)
        # print(nl)
    cc = 0
    for row in nl:
        for col in row:
            if col=="#":
                cc+=1
    return cc


if __name__ == "__main__":
    # print(f"{one(ls) = }") # 
    print(f"{two(ls) = }") # 

File: 11.cpp
#include "aoc.hpp"

const string inputname = "11.in";
using sol_t = long long;
using joltage_t = long long;
using seat_map_t = vector<string>;

seat_map_t seat_map{{}};

sol_t one(seat_map_t lin)
{

    bool changeflag = true;
    auto cidc = 0;
    seat_map_t nl{};
    seat_map_t ls = lin;
    while (changeflag)
    {
        cidc += 1;
        cout << cidc << '\n';
        changeflag = false;
        auto ir = 0;
        auto ic = 0;
        nl = {};
        for (auto & row:ls)
        {
            auto r = "";
            ic = 0;
            for(auto & col: row)
            {
                if (*col=="L")
                {
                    auto count = 0;
                    vector<int> s{-1,0,1};
                    for (auto &dc: s)
                    {
                        for (auto &dr: s)
                        {
                            if ((ic+dc)==row.size()), break;
                            if ((ir+dr)==ls.size()), break;
                            if ((ir+dr)==-1), break;
                            if ((ic+dc)==-1), break;
                            if ((dc==dr) &&(dr==0)), break;
                            if (ls[ir+dr][ic+dc] == "#")
                            {
                                count+=1;
                            }
                        }
                    }
                    if (count==0)
                    {
                        r+="#";
                        changeflag = true;
                    }
                    else
                    {
                        r+="L"
                    }
                }
                else if (col=="#")
                {
                    auto count = 0;
                    vector<int> s{-1,0,1};
                    for (auto &dc: s)
                    {
                        for (auto &dr: s)
                        {
                            if (ic+dc)==len(row), break;
                            if (ir+dr)==len(ls), break;
                            if (ir+dr)==-1, break;
                            if (ic+dc)==-1, break;
                            if (dc==dr) &&(dr==0), break;
                            if ls[ir+dr][ic+dc] == "#"
                            {
                                count+=1;
                            }
                        }
                    }
                    if (count>=4)
                    {
                        r+="L";
                        changeflag = true;
                    }
                    else
                    {
                        r+="#";
                    }
                }
                else
                {
                    r+=ls[ir][ic];
                }
                ic+=1;
            }
            nl.push_back(r);
            ir+=1;
        }
        ls = nl;
    }
    auto cc = 0;
    for (auto & row: nl)
    {
        for (auto & col: row)
        {
            if (col == '#')
            {
                ++cc;
            }
        }
    }
    return cc;
}

sol_t two(seat_map_t lin)
{
    return 0;
}

int main()
{
    ifstream infile(inputname);
    string s;
    string chars = "\n\r";
    while( getline( infile, s ) ) // for each line read from the file
    {
        for (char c: chars)
        {
            s.erase(remove(s.begin(), s.end(), c), s.end());
        }
        seat_map.push_back(s);
    }
    cout << "one = " << one(seat_map) << '\n'; // 
    cout << "two = " << two(seat_map) << '\n'; // 
    return 0;
}